controller/src/main/java/org/jetbrains/MavenController.java

```diff
@@ -62,42 +105,66 @@ private void exec(InvocationRequest request) throws IOException {
                 throw new IllegalStateException("Build failed.");
             }
             log.info("Invoker execution successful");
+            exitCode = (result.getExitCode() == 0) ? StatusCode.OK : StatusCode.ERROR;
         } catch (Exception e) {
-            // Signals an error during the construction of the command line used to invoke Maven
-            // e.g. illegal invocation arguments.
             log.error(e.getMessage());
         }
+        return exitCode;
     }
 
-    public void compile() throws IOException {
-        log.info("MavenController::compile started");
-        compile(false);
-        log.info("MavenController::compile ended");
+    public StatusCode compile(String projectDirectory) {
+        return compile(projectDirectory, false);
     }
-
-    public void compile(boolean clean) throws IOException {
+    public StatusCode compile(String projectDirectory, boolean clean) {
+        log.info("MavenController::compile started");
         InvocationRequest request = new DefaultInvocationRequest();
         if (clean) {
             request.setGoals(Arrays.asList("clean", "compile"));
         } else {
             request.setGoals(Collections.singletonList("compile"));
         }
-        exec(request);
-    }
+        request.setBaseDirectory(new File (projectDirectory));
 
-    public void install() throws IOException {
-        log.info("MavenController::install started");
-        install(false);
-        log.info("MavenController::install ended");
+        StatusCode ret = exec(request);
+        log.info("MavenController::compile ended");
+        return ret;
     }
 
-    public void install(boolean clean) throws IOException {
+    public StatusCode install(String projectDirectory) {
+        return install(projectDirectory, false);
+    }
+    public StatusCode install(String projectDirectory, boolean clean) {
+        log.info("MavenController::install started");
         InvocationRequest request = new DefaultInvocationRequest();
         if (clean) {
             request.setGoals(Arrays.asList("clean", "install"));
         } else {
             request.setGoals(Collections.singletonList("install"));
         }
-        exec(request);
+        request.setBaseDirectory(new File (projectDirectory));
+        StatusCode ret = exec(request);
+        log.info("MavenController::install ended");
+        return ret;
+    }
+
+    public StatusCode compileTargets(CompileParams compileParams) {
+        StatusCode resultCode = StatusCode.CANCELLED;
+
+        for(BuildTargetIdentifier x: compileParams.getTargets()) {
+            String path = x.getUri();
```

Comment:
`getUri` doesn't return path but URI, thereof:

```suggestion
            String path = URI.create(x.getUri()).getPath();
```